rafthttp/stream.go

```diff
@@ -0,0 +1,388 @@
+	return &streamDialer{
+		u:       u,
+		tr:      tr,
+		from:    from,
+		to:      to,
+		cid:     cid,
+		t:       t,
+		updatec: make(chan struct{}, 1),
+	}
+}
+
+func (d *streamDialer) dial() (*incomingStream, error) {
+	d.mu.RLock()
+	u := d.u
+	term := d.msgAppTerm
+	d.mu.RUnlock()
+
+	uu, err := url.Parse(u)
+	if err != nil {
+		return nil, fmt.Errorf("parse url %s error: %v", d.u, err)
+	}
+	uu.Path = path.Join(RaftStreamPrefix, string(d.t), d.from.String())
+	req, err := http.NewRequest("GET", uu.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("new request to %s error: %v", d.u, err)
+	}
+	req.Header.Set("X-Etcd-Cluster-ID", d.cid.String())
+	req.Header.Set("X-Raft-To", d.to.String())
+	switch d.t {
+	case streamTypeMsgApp:
+		req.Header.Set("X-Raft-Term", strconv.FormatUint(term, 10))
+	case streamTypeGeneral:
+	default:
+		log.Panicf("rafthttp: unhandled stream type %s", d.t)
+	}
+	resp, err := d.tr.RoundTrip(req)
+	if err != nil {
+		return nil, fmt.Errorf("error posting to %q: %v", d.u, err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		resp.Body.Close()
+		return nil, fmt.Errorf("unhandled http status %d", resp.StatusCode)
+	}
+
+	stream := &incomingStream{}
+	switch d.t {
+	case streamTypeMsgApp:
+		stream.decoder = &msgAppDecoder{r: resp.Body, local: d.from, remote: d.to, term: term}
+	case streamTypeGeneral:
+		stream.decoder = &generalDecoder{r: resp.Body}
+	default:
+		log.Panicf("rafthttp: unhandled stream type %s", d.t)
+	}
+	stream.Closer = resp.Body
+	return stream, nil
+}
+
+func (d *streamDialer) update(u string) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.u = u
+	select {
+	case d.updatec <- struct{}{}:
+	default:
+	}
+}
+
+func (d *streamDialer) updateMsgAppTerm(term uint64) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.msgAppTerm == term {
+		return
+	}
+	d.msgAppTerm = term
+	select {
+	case d.updatec <- struct{}{}:
+	default:
+	}
+}
+
+type streamReader struct {
+	dialer *streamDialer
+	recvc  chan<- raftpb.Message
+	stopc  chan struct{}
+	done   chan struct{}
+}
+
+func startStreamReader(d *streamDialer, recvc chan<- raftpb.Message) *streamReader {
+	r := &streamReader{
+		dialer: d,
+		recvc:  recvc,
+		stopc:  make(chan struct{}),
+		done:   make(chan struct{}),
+	}
+	go r.run()
+	return r
+}
+
+func (cr *streamReader) run() {
+	var stream *incomingStream
+	var err error
+	reset := func() {
+		if stream != nil {
+			stream.Close()
+			stream = nil
+		}
+	}
+	for {
+		select {
+		case <-cr.dialer.updatec:
+			reset()
+		case <-cr.stopc:
+			reset()
+			close(cr.done)
+			return
+		default:
+		}
+		if stream == nil {
+			stream, err = cr.dialer.dial()
+			if err != nil {
+				log.Printf("rafthttp: dial error: %v", err)
+				continue
+			}
+		}
+		m, err := stream.decode()
+		switch {
+		case err != nil && err != io.EOF:
+			log.Printf("rafthttp: encountered error reading the connection: %v", err)
+			fallthrough
+		case err != nil:
+			reset()
+		case isHeartbeatMessage(m):
```

Comment:
comment // do nothing for heartbeat, TODO: record the recent activity of remote peer